enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,574 @@
+
+##### Usage of the json representation of packages by olm components
+
+The registry database that is created inside the indexes are used by olm components to query the state of the index(eg to list packages in the index, to query for the list of APIs required by an operator during installation of that operator etc). The representation of packages as json provides for an unique opportunity for olm components to work with json representations instead of the sqllite database, operations on which has proven to be costly in many instances. To allow for the eventual phase out of using the sqllite database, a package representation tailored for olm components will be created alongside the package representation that index/operator authors will interact with. The properties on this package json will be aligned to properties used by olm components like dependency resolver and their values will mirror the values from the represenation for index/operator authors. It will also have additional information per bundle required by these components. For example, when a olm component queries for a package in the index, it will be searching for a file named `olm.<package-name>.json`. An example of the possible content of the file is enumerated next: 
+
+```bash 
+$ cat olm.etcd.json
+"etcd": {
+        "olm.package": "etcd",
+        "bundles": [
+            {
+                "olm.bundleImage": "quay.io/random/etcdv0.0.1",
+                "olm.packageVersion": "v0.0.1",
+                "channels": ["alpha"],
+                 "ownedAPIs": [
+                    {
+                        "group": "etcd.database.coreos.com",
+                        "version": "v1alpha1",
+                        "kind": "etcd"
+                    }
+                ],
+                "requiredAPIs": []
+            },
+            {
+                "olm.bundleImage": "quay.io/somethingelse/etcdv0.0.2",
+                "olm.packageVersion": "v0.0.2",
+                "channels": ["alpha"],
+                "ownedAPIs": [
+                    {
+                        "group": "etcd.database.coreos.com",
+                        "version": "v1alpha1",
+                        "kind": "etcd"
+                    }
+                ],
+                "requiredAPIs": []
+            }
+        ],
+        "defaultChannel": "alpha",
+        "icon": "asdfasdfoipuasdfvasdfqer",
+        "channelUpgradeGraphs": {
+            "channels": [
+                {
+                    "name": "alpha",
+                    "upgrades": [
```

Comment:
how are skips+skipranges represented?

Reply:
@bparees when a bundle mentions skips/skipranges, doesn't the `channel_entry` table in the registry database contain the information about which bundle that particular bundle will upgrade to? 
That information will be captured in the `channelUpgradeGraphs` section the same way it will when a bundle is added in with `replaces` information. I don't see why this may stray from `replaces` or any other mode, unless I'm missing something else.

Reply:
if i understand you correctly, then in order for bundle C to "skip" bundles A and B, i must explicitly add the appropriate upgrades to/from entries to C, A, and B.

It's not sufficient for me to state that C skips B and B skips A, i have to explicitly declare that "C upgrades from A+B", that "B upgrades to C" and "A upgrades to C".

correct?

If so, that seems to put an additional burden on the author since they need to understand the entire upgrade chain so they can explicitly specify the to/from edges correctly on each new node, because no edges are implied.

Reply:
>if i understand you correctly, then in order for bundle C to "skip" bundles A and B, i must explicitly add the appropriate upgrades to/from entries to C, A, and B.
>
>It's not sufficient for me to state that C skips B and B skips A, i have to explicitly declare that "C upgrades from A+B", that "B upgrades to C" and "A upgrades to C".

I am a little confused by your example so let me try to use the following example to make things clear and address your concerns: 
A channel in a package has bundles A and B where B was added to replace A. That information is captured in the `channel_entry` table in the following way: 

```
entry id   |    channel_name   |  operatorbunle_name   | replaces 
1                    alpha               A 
2                    alpha               B                   1
```
The upgrade graph in the config for this channel would look like so: 

```yaml
"channelUpgradeGraphs": {
            "channels": [
                {
                    "name": "alpha",
                    "upgrades": [
                        {
                            "to": {
                                "bundle": A
                            }
                        },
                        {
                            "from": {
                                "bundles": A
                            },
                            "to": {
                                "bundle": B
                            }
                        }
                    ]
                }
            ]
        }
```

A new bundle C is added that `skips` B and replaces A+B. That information is captured in the `channel_entry` table in the following way: 

```
entry id   |    channel_name   |  operatorbunle_name   | replaces 
1                    alpha               A                
2                    alpha               B                   1
3                    alpha               C                   1 
4                    alpha               C                   2 
```

The graph in the config file for which will look like: 

```yaml
"channelUpgradeGraphs": {
            "channels": [
                {
                    "name": "alpha",
                    "upgrades": [
                        {
                            "to": {
                                "bundle": A
                            }
                        },
                        {
                            "from": {
                                "bundles": A
                            },
                            "to": {
                                "bundle": B
                            }
                        },
                        {
                            "from": {
                                "bundles": A,B
                            },
                            "to": {
                                "bundle": C
                            }
                        },
                    ]
                }
            ]
        }
```
Which means whether the dependency resolver queries the db or the config file, A will be upgraded to C, B will be upgraded to C, and B can only be installed if the subscription for that package mentions B as the starting CSV. B will not be in the upgrade path from A. 

Basically, the like every other information in the config file, the upgrade path info will also be a mirror of the info in the db.
 
 >If so, that seems to put an additional burden on the author since they need to understand the entire upgrade chain so they can explicitly specify the to/from edges correctly on each new node, because no edges are implied.

As for this point, the config file doesn't add any more complexity to the UX than the db would have if they could interact directly with the db. 

I see the potential for a confusing UX here, and hopefully a command to visualize the upgrade graph that'll be proposed [here](https://github.com/operator-framework/enhancements/pull/26) can help alleviate that somewhat. But this is exactly the kind of use case that can serve as a motivation for rewriting how the graph is represented and how the dependency resolver uses that graph if needed, once the db is no longer in use. 
cc: @dmesser

Reply:
This makes sense to me. What I like to call out is how this would look like in general when we talk about `semver` mode. With this mode we wanted to alleviate users from having to deal with these bundle "to" and "from" references in general and just let `semver` to do the ordering.

Reply:
taking this to slack: https://coreos.slack.com/archives/C3VS0LV41/p1610646052005300